File: Data Structures & Algorithms/climbing-stairs/submission-2.py
class Solution:
    def climbStairs(self, n: int) -> int:

        memo = {}
        
        def climb(step):

            if step == n:
                return 1
            
            if step > n:
                return 0
            
            if step in memo:
                return memo[step]

            memo[step] = climb(step + 1) + climb(step + 2)

            return memo[step]
        
        return climb(0)

File: Data Structures & Algorithms/n-th-tribonacci-number/submission-2.cpp
using namespace std;

class Solution {
public:
    int tribonacci(int n) {
        
        unordered_map<int, int> tab = {{0, 0,}, {1, 1}, {2, 1}};

        for (int step = 3; step <= n; step++) {

            int fib1 = tab.count(step - 1) ? tab[step - 1] : 0;
            int fib2 = tab.count(step - 2) ? tab[step - 2] : 0;
            int fib3 = tab.count(step - 3) ? tab[step - 3] : 0;

            int fib = fib1 + fib2 + fib3;

            tab[step] = fib;
        }

        return tab[n];

    }
};